### ISSUE TITLE
lfe_io_pretty:term/1 is not so pretty for (map k0 v0 ... kn vn)

### ISSUE THREAD

**Issue by yurrriq** [lfe/lfe#183]
I'm mostly leaving this here as a note to self. I'd like to take a stab at this when I get time, unless someone else wants to beat me to it.

Real world example output:

``` lfe
(map
 #"id"
 id
 #"name"
 name
 #"known_by"
 known_by
 #"bio"
 bio
 #"image"
 image
 #"topics"
 topics
 #"links"
 links)
```

`#m(....)` forms are handed differently, if not preferably:

``` lfe
`#M(#"date_modified" ,date_modified
    #"date_published" ,date_published #"id" ,id #"masters" ,masters
    #"sections" ,sections #"slug" ,slug #"summary" ,summary
    #"thumbnail" ,thumbnail #"title" ,title)
```

It'd be cool to align keys and/or values, but it seems like that might easily become a fool's errand. What's more, that style is probably a matter of personal (and subjective) taste.

All things considered, I'd be fine with `(map ...)` forms being printed as `#m(...)` ones are currently, potentially with the precondition that `(> (length form) 2)`, so as not to mess up any `*:map/2` forms.

**Comment by rvirding** [comment]
No, there is no special handling of the (map ...) form, it is the same as for lists and tuples. The special handling for the #m(...) literal is quite simple: it tries to keep the key and value on the same and introduces line breaks when this is not possible. It would be easy to fix the (map ...) so it behaves in the same way as for the map literal.

I think trying to put too much work into indenting the map would not be worth it.

There would be no problems with *:map/2 forms unless they were local calls but with 2 args it would look quite normal. Only question is whether you would prefer:

``` lisp
(map #"id" id
     #"name" name
     #"known-by" known-by)
```

or

``` lisp
(map
  #"id" id
  #"name" name
  #"known-by" known-by)
```

:-)

**Comment by yurrriq** [comment]
> I think trying to put too much work into indenting the map would not be worth it.

I agree wholeheartedly.

> Only question is whether you would prefer...

Personally, I prefer the former.

**Comment by yurrriq** [comment]
@rvirding, any chance you could implement this:

``` lfe
(map #"id" id
     #"name" name
     #"known-by" known-by)
```

If you point me in the right direction, I could make a PR instead.

**Comment by rvirding** [comment]
Yes, I have been thinking about as this one is a little tricky. My plan is to make it behave in basically the same way as for literal maps, ans preferably to reuse most of the code. Try a PR if you wish and will where we end up.

**Comment by yurrriq** [comment]
I did a [hackish first pass](https://github.com/yurrriq/lfe/commit/f571dc3dacd90312ed137652cdd5faa12ac71a68) with minimal code reuse. It worked pretty well in the few tests I ran. If we want to go with this approach, it seems like we ought to abstract out `F` (for normal maps it's `maps:to_list/1`) and turn `map/4` into `map/5` with the fifth parameter being `F`. Does that make sense to anyone but me?
